resolver/resolver.go

```diff
@@ -407,7 +407,37 @@ func isUDP(w dns.ResponseWriter) bool {
 // permitted length of the given transmission channel.
 // See https://tools.ietf.org/html/rfc1035#section-4.2.1
 func truncate(m *dns.Msg, udp bool) *dns.Msg {
-	m.Truncated = udp && m.Len() > dns.MinMsgSize
+	bufsize := 0
+	switch udp {
```

Comment:
Like this? It's pretty much what you suggested but without putting the assigments in the if clause. IMO that should be only done if you want the variable to be in the scope of the condition only.
(I also think that small, non-edns udp responses will always get truncated because max == 0 after line 7 in your suggestion, right?)